linux: Fix multiple keyring prompts

jkful opened the pull request:
Related to #17460

On Linux, multiple keyring unlock prompts could appear due to `keyring.unlock()` calls. This change:

- Checks `item.is_locked()` before calling `item.unlock()` in read/delete
- Only unlocks collection in write if `IsLocked` error is returned

Users should now see at most one prompt if keyring is locked.

Tested on Debian 13.2 (trixie) with GNOME Keyring.

Release Notes:

- Fixed multiple keyring unlock prompts on Linux

jkful commented:
@cla-bot check

bilelmoussaoui commented:
The changes looks good to me. Although, keep in mind, nothing in the spec says that you can always search for items even if the collection where they are stored is locked. Would be nice if people could verify if this assumption would work also in kwallet/keepass

probably-neb commented:
Hey sorry for the delay on this. This change makes a lot of sense. Once the merge conflicts are resolved I'd be happy to merge

jkful commented:
@probably-neb No worries, just got around to this again. Should be all set here, let me know if anything else is needed.

DaniPopes commented:
@smitbarmase @probably-neb Hello, is there anything left to do here? Would love to have this in. Spawning multiple agent sessions in Zed on Linux exposes a bug in Gnome Keyring daemon which makes it crash and logs out the keyring completely: https://github.com/zed-industries/zed/issues/17460#issuecomment-4406107125

jkful committed:
linux: Fix multiple keyring prompts

On Linux, multiple keyring unlock prompts could appear due to
keyring.unlock() calls. This change:

* Checks item.is_locked() before calling item.unlock() in read/delete
* Only unlocks the collection in write if an IsLocked error is returned
* Caches oo7::Keyring in a OnceCell so all credential operations share a
  single D-Bus connection, preventing gnome-keyring-daemon crashes from
  concurrent connections

Tested on Debian 13.2 (trixie) with GNOME Keyring.

Release Notes:

* Fixed multiple keyring unlock prompts on Linux

jkful commented:
@cla-bot check

probably-neb commented on crates/gpui_linux/src/linux/platform.rs:
```diff
@@ -746,6 +766,22 @@ impl<P: LinuxClient + 'static> Platform for LinuxPlatform<P> {
+    keyring: &Arc<OnceCell<Result<Arc<oo7::Keyring>>>>,
+) -> Result<Arc<oo7::Keyring>> {
+    keyring
+        .get_or_init(|| async {
```
```suggestion
        .get_or_try_init(|| async {
```

We should probably use the `OnceCell::get_or_try_init` api to leave the cell uninitialized if we hit a transient load error so that we don't perma-block ourselves from using the keyring if the first request fails (transiently or otherwise).

probably-neb commented on crates/gpui_linux/src/linux/platform.rs:
```diff
@@ -653,36 +661,46 @@ impl<P: LinuxClient + 'static> Platform for LinuxPlatform<P> {
             for item in items.into_iter() {
                 if item.label().await.is_ok_and(|label| label == KEYRING_LABEL) {
+                    if item.is_locked().await? {
+                        item.unlock().await?;
```
This seems potentially racy, multiple callers could hit `is_locked() == true` before a single `unlock()` resolves. It's probably fine to serialize these calls across different callers
